README.md
# gopop - Wrapper arround the Poppler CLI

This is a wrapper around the [Poppler CLI tools](https://poppler.freedesktop.org/).
It is mostly a toy project to develop something open source.
Feel free to contribute or use it but don't expect high availability support or maintenance
(at least for the time being).

The use case of this library is probably to integrate Poppler into a larger application.
To archive wrapping Poppler commands, this library also implements the required parsing / util code which
you can also you if you find any of it useful.

## Versioning

This project will not maintain a stable API until v1.0.0 is released.

## Why wrap the CLI?

There are [a couple of modules integrating Poppler via C bindings](https://pkg.go.dev/search?q=poppler&m=).
This project aims to be an alternative that makes use of parsing the CLI output of Poppler commands.
This may not be the most advanced approach but it has the advantage of being cross-platform and
generally easier to implement (at least for me).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- First working parser for `pdfinfo` cmd output.

<!-- TODO: Update these once the first versions are tagged -->

[unreleased]: https://github.com/FabianTe/gopop/compare/v0.0.2...HEAD
[0.0.2]: https://github.com/FabianTe/gopop/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/FabianTe/gopop/releases/tag/v0.0.1


pdfinfo_test.go
package gopop

import (
	"bytes"
	"encoding/json"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"testing"
)

func TestPDFInfoParser_Parse(t *testing.T) {
	in := `Producer:        PDFKit.NET 23.1.101.39834 DMV9
CreationDate:    Wed Apr 19 12:58:37 2023 CEST
ModDate:         Wed Apr 19 12:58:37 2023 CEST
Custom Metadata: no
Metadata Stream: yes
Tagged:          no
UserProperties:  no
Suspects:        no
Form:            AcroForm
JavaScript:      no
Pages:           2
Encrypted:       no
Page    1 size:  595.32 x 841.92 pts (A4)
Page    1 rot:   0
Page    2 size:  595.32 x 841.92 pts (A4)
Page    2 rot:   0
File size:       230296 bytes
Optimized:       no
PDF version:     1.7
`
	parser := NewPDFInfoParser(bytes.NewBufferString(in))
	info, err := parser.Parse()
	require.NoError(t, err)
	require.NotNil(t, info)

	assert.Len(t, info.Pages, 2, "wrong amount of pages")
	assert.Equal(t, "AcroForm", info.Properties["Form"])
	assert.Equal(t, "2", info.Properties["Pages"])
	assert.Equal(t, "230296 bytes", info.Properties["File size"])

	t.Run("json representation", func(t *testing.T) {

		j, err := json.Marshal(info)
		require.NoError(t, err)

		assert.JSONEq(t, `{
			"Properties": {
				"CreationDate": "Wed Apr 19 12:58:37 2023 CEST",
				"Custom Metadata": "no",
				"Encrypted": "no",
				"File size": "230296 bytes",
				"Form": "AcroForm",
				"JavaScript": "no",
				"Metadata Stream": "yes",
				"ModDate": "Wed Apr 19 12:58:37 2023 CEST",
				"Optimized": "no",
				"PDF version": "1.7",
				"Pages": "2",	
				"Producer": "PDFKit.NET 23.1.101.39834 DMV9",
				"Suspects": "no",
				"Tagged": "no",
				"UserProperties": "no"
			},
			"Pages": {
				"1": {
					"rot": "0",
					"size": "595.32 x 841.92 pts (A4)"
				},
				"2": {
					"rot": "0",
					"size": "595.32 x 841.92 pts (A4)"
				}
			}
		}`, string(j))
	})

	t.Run("info methods", func(t *testing.T) {
		assert.Equal(t, "PDFKit.NET 23.1.101.39834 DMV9", info.Producer())
	})

	t.Run("page methods", func(t *testing.T) {
		page1 := info.Page(1)
		require.NotNil(t, page1)

		t.Run("raw values", func(t *testing.T) {
			assert.Equal(t, "595.32 x 841.92 pts (A4)", page1.SizeRaw())
			assert.Equal(t, "0", page1.RotRaw())
		})

		t.Run("parsed values", func(t *testing.T) {
			size := page1.Size()
			assert.Equal(t, 595.32, size.Width)
			assert.Equal(t, 841.92, size.Height)
			assert.Equal(t, "pts", size.Unit)
			assert.Equal(t, "(A4)", size.Note)
		})
	})
}


pdfinfo.go
package gopop

import (
	"bufio"
	"fmt"
	"io"
	"regexp"
	"strconv"
	"strings"
)

type PageSize struct {
	Width  float64
	Height float64
	Unit   string
	Note   string
}

// PDFInfoPage contains information about a single page.
//
//	Page    1 size:  595.32 x 841.92 pts (A4)
//	Page    1 rot:   0
//	Page    2 size:  595.32 x 841.92 pts (A4)
//	Page    2 rot:   0
type PDFInfoPage map[string]string

var regexPDFInfoPageSize = regexp.MustCompile("^(.*) x (.*) (pts) ?(.*)$")

// Size parses the value of SizeRaw and returns a PageSize.
//
// Panics if parsing fails
func (page PDFInfoPage) Size() PageSize {
	raw := page.SizeRaw()
	find := regexPDFInfoPageSize.FindStringSubmatch(raw)
	if find == nil {
		panic(fmt.Errorf("parsing page size %q failed", raw))
	}
	w := find[1]
	h := find[2]
	unit := find[3]
	note := ""
	if len(find) > 4 {
		note = find[4]
	}

	wf, err := strconv.ParseFloat(w, 64)
	if err != nil {
		panic(fmt.Errorf("parsing %q to float failed", w))
	}

	hf, err := strconv.ParseFloat(h, 64)
	if err != nil {
		panic(fmt.Errorf("parsing %q to float failed", h))
	}

	return PageSize{
		Width:  wf,
		Height: hf,
		Unit:   unit,
		Note:   note,
	}
}

func (page PDFInfoPage) SizeRaw() string {
	return page["size"]
}

func (page PDFInfoPage) RotRaw() string {
	return page["rot"]
}

type PDFInfo struct {
	// Properties holds arbitrary key/value pairs that are not further interpreted.
	// An example:
	//
	//     Producer:        PDFKit.NET 23.1.101.39834 DMV9
	//     CreationDate:    Wed Apr 19 12:58:37 2023 CEST
	//     ModDate:         Wed Apr 19 12:58:37 2023 CEST
	//     Custom Metadata: no
	//     Metadata Stream: yes
	//     Tagged:          no
	//     UserProperties:  no
	//     Suspects:        no
	//     Form:            AcroForm
	//     JavaScript:      no
	//     Pages:           2
	//     Encrypted:       no
	//     File size:       230296 bytes
	//     Optimized:       no
	//     PDF version:     1.7
	Properties map[string]string

	Pages map[string]PDFInfoPage
}

func NewPDFInfo() *PDFInfo {
	return &PDFInfo{
		Properties: map[string]string{},
		Pages:      map[string]PDFInfoPage{},
	}
}

func (info *PDFInfo) Producer() string {
	return info.Properties["Producer"]
}

// editPage is a helper method to edit a page. This method will create a page if
// it does not exist.
func (info *PDFInfo) editPage(page string, edit func(page PDFInfoPage)) {
	p := info.Pages[page]
	if p == nil {
		p = PDFInfoPage{}
		info.Pages[page] = p
	}
	edit(p)
}

// Page returns the given page or nil if it does not exist.
func (info *PDFInfo) Page(page int) PDFInfoPage {
	return info.Pages[strconv.Itoa(page)]
}

// PDFInfoParser is a parser for the output of the pdfinfo command.
//
// This parser is not fully tested and may break if pdfinfo outputs something that we don't expect.
type PDFInfoParser struct {
	reader io.Reader
	info   *PDFInfo
}

var (
	regexParseLine    = regexp.MustCompile("^(.*?):[\t ]*(.*)$")
	regexParsePageKey = regexp.MustCompile("^Page[\\t ]*?(\\d+)[\\t ](.*)$")
)

func NewPDFInfoParser(reader io.Reader) *PDFInfoParser {
	return &PDFInfoParser{reader: reader, info: NewPDFInfo()}
}

func (parser *PDFInfoParser) Parse() (*PDFInfo, error) {
	scanner := bufio.NewScanner(parser.reader)

	for scanner.Scan() {
		line := scanner.Text()
		if err := parser.parseLine(line); err != nil {
			return nil, err
		}
	}

	return parser.info, nil
}

func (parser *PDFInfoParser) parseLine(line string) error {
	match := regexParseLine.FindStringSubmatch(line)
	if match == nil {
		return nil
	}

	key := match[1]
	value := match[2]

	switch {
	case strings.HasPrefix(key, "Page "):
		if err := parser.parsePageKey(key, value); err != nil {
			return err
		}
		break
	default:
		parser.info.Properties[key] = value
		break
	}

	return nil
}

func (parser *PDFInfoParser) parsePageKey(key string, value string) error {
	keyMatch := regexParsePageKey.FindStringSubmatch(key)
	if keyMatch == nil {
		return fmt.Errorf("invalid page key %q", key)
	}

	pageNumber := keyMatch[1]
	pageKey := keyMatch[2]

	parser.info.editPage(pageNumber, func(page PDFInfoPage) {
		page[pageKey] = value
	})

	return nil
}
